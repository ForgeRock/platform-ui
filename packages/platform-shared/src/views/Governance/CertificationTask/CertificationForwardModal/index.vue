<!-- Copyright (c) 2023 ForgeRock. All rights reserved.

This software may be modified and distributed under the terms
of the MIT license. See the LICENSE file for details. -->
<template>
  <ValidationObserver
    v-slot="{ invalid }"
    ref="observer">
    <BModal
      :id="modalId"
      :ok-disabled="invalid"
      :ok-title="$t('governance.certificationTask.actionsModal.forwardItem')"
      :title="$t('governance.certificationTask.actionsModal.forwardItem')"
      :static="isTesting"
      @ok="forwardItem"
      @hidden="reset"
      cancel-variant="link"
      ok-variant="info"
      size="lg">
      <p>
        {{ $t('governance.certificationTask.actionsModal.forwardItemDescription') }}:
      </p>
      <div class="mb-4">
        <BFormRadioGroup
          class="mb-4"
          stacked
          v-model="forwardToType"
          :options="forwardToOptions"
        />
        <FrGovResourceSelect
          v-model="forwardToResource"
          :resource-path="isUserSelected ? 'user' : 'role'"
          :label="$t('governance.certificationTask.actionsModal.forwardTo')"
          validation="required"
        />
      </div>

      <!-- comment -->
      <FrField
        v-model="comment"
        class="mb-4"
        type="textarea"
        validation="required"
        :label="$t('governance.certificationTask.actionsModal.comment')" />
    </BModal>
  </ValidationObserver>
</template>

<script>
import {
  BFormRadioGroup,
  BModal,
} from 'bootstrap-vue';
import { ValidationObserver } from 'vee-validate';
import FrField from '@forgerock/platform-shared/src/components/Field';
import FrGovResourceSelect from '@forgerock/platform-shared/src/components/governance/GovResourceSelect';

/**
 * @constant {Object}
 * @default
 */
const FORWARD_TO_TYPES = {
  role: 'role',
  user: 'user',
};

export default {
  name: 'CertificationForwardModal',
  components: {
    BFormRadioGroup,
    BModal,
    FrField,
    FrGovResourceSelect,
    ValidationObserver,
  },
  props: {
    id: {
      type: String,
      default: '',
    },
    bulk: {
      type: Boolean,
      default: false,
    },
    modalId: {
      type: String,
      default: 'CertificationTaskForwardAccountModal',
    },
  },
  data() {
    return {
      comment: '',
      forwardToUser: '',
      forwardToRole: '',
      forwardToType: FORWARD_TO_TYPES.user,
      forwardToOptions: [
        { text: this.$t('governance.certificationTask.actionsModal.toUser'), value: FORWARD_TO_TYPES.user },
        { text: this.$t('governance.certificationTask.actionsModal.toRole'), value: FORWARD_TO_TYPES.role },
      ],
      isTesting: false,
    };
  },
  methods: {
    forwardItem() {
      const payload = {
        id: this.bulk ? undefined : this.id,
        comment: this.comment,
        newActorId: this.forwardToType === FORWARD_TO_TYPES.user
          ? this.forwardToUser
          : this.forwardToRole,
      };
      const emitter = this.bulk ? 'forward-bulk' : 'forward-item';
      this.$emit(emitter, payload);
    },
    reset() {
      this.comment = '';
      this.forwardToUser = '';
      this.forwardToRole = '';
    },
  },
  computed: {
    isUserSelected() {
      return this.forwardToType === 'user';
    },
    forwardToResource: {
      get() {
        return this.isUserSelected ? this.forwardToUser : this.forwardToRole;
      },
      set(value) {
        if (this.isUserSelected) {
          this.forwardToUser = value;
        } else {
          this.forwardToRole = value;
        }
      },
    },
  },
};
</script>

<style lang="scss" scoped>
.input-line-height {
  line-height: 1.2;
}
</style>
