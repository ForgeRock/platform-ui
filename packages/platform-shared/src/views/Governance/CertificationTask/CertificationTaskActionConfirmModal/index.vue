<!-- Copyright (c) 2023 ForgeRock. All rights reserved.

This software may be modified and distributed under the terms
of the MIT license. See the LICENSE file for details. -->
<template>
  <BModal
    :id="modalId"
    size="lg"
    ok-variant="primary"
    cancel-variant="link"
    :ok-title="$t(`governance.certificationTask.actionsModal.${modalOptions.okLabel}`)"
    @ok="modalOptions.okFunction(confirmMessage)"
    :title="$t(`governance.certificationTask.actionsModal.${modalOptions.title}`)"
    :static="isTesting"
    @hidden="reset">
    <div class="modal-container">
      <span>{{ $t(`governance.certificationTask.actionsModal.${modalOptions.description}`) }}</span>
      <FrField
        v-model="confirmMessage"
        class="mb-4 mt-4"
        type="textarea"
        :label=" $t(`governance.certificationTask.actionsModal.${modalOptions.placeHolder}`)"
        :rows="4"
        :max-rows="4"
      />
    </div>
  </BModal>
</template>

<script>
import {
  BModal,
} from 'bootstrap-vue';
import FrField from '@forgerock/platform-shared/src/components/Field';

export default {
  name: 'CertificationTaskActionConfirmModal',
  components: {
    BModal,
    FrField,
  },
  data() {
    return {
      confirmMessage: '',
      isTesting: false,
    };
  },
  props: {
    modalOptions: {
      type: Object,
      default: () => ({}),
    },
    modalId: {
      type: String,
      default: 'CertificationTaskActionConfirmAccountModal',
    },
  },
  methods: {
    reset() {
      this.confirmMessage = '';
    },
  },
};
</script>
