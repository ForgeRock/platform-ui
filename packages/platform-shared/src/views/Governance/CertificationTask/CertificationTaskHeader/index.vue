<!-- Copyright (c) 2023 ForgeRock. All rights reserved.

This software may be modified and distributed under the terms
of the MIT license. See the LICENSE file for details. -->
<template>
  <FrNavbar
    hide-dropdown
    hide-toggle
    :show-docs-link="false"
    :show-help-link="false"
    :show-profile-link="false">
    <template #center-content>
      <h4>
        <span>{{ campaignDetails.name }}</span>
      </h4>
    </template>
    <template
      v-if="!hideSignOff"
      #right-content>
      <div
        v-if="!isSaving && !firstLoad"
        class="d-flex m-2">
        <FrIcon
          class="mr-2 color-green"
          name="check" />
        <span class="text-muted">
          {{ $t('governance.certificationTask.allChangesSaved') }}
        </span>
      </div>
      <FrSpinner
        v-if="isSaving"
        :button-spinner="isSaving"
        :name="$t('governance.certificationTask.isSaving')"
        class="mr-3 text-primary"
        size="sm" />
      <div
        v-if="isSaving"
        class="media-body mr-3"
        v-html="$t('governance.certificationTask.isSaving')" />
      <BButton
        variant="primary"
        :disabled="disableSignoff"
        @click="$emit('sign-off')">
        {{ $t('governance.certificationTask.signOff') }}
      </BButton>
      <BDropdown
        v-if="campaignDetails.enableForward && campaignDetails.allowBulkCertify"
        class="ml-2"
        variant="link"
        boundary="window"
        no-caret
        :disabled="isSaving">
        <template #button-content>
          <FrIcon
            class="text-dark md-24"
            name="more_horiz" />
        </template>
        <BDropdownItem
          @click="onForwardAllTasks">
          <FrIcon
            class="mr-2"
            name="redo" />
          {{ $t('governance.certificationTask.actions.forward') }}
        </BDropdownItem>
      </BDropdown>
    </template>
  </FrNavbar>
</template>

<script>
import {
  BButton,
  BDropdown,
  BDropdownItem,
} from 'bootstrap-vue';
import FrNavbar from '@forgerock/platform-shared/src/components/Navbar/';
import FrIcon from '@forgerock/platform-shared/src/components/Icon';
import FrSpinner from '@forgerock/platform-shared/src/components/Spinner/';
import { forwardCertificationTasks } from '@forgerock/platform-shared/src/api/governance/CertificationApi';

export default {
  name: 'CertificationTaskHeader',
  components: {
    FrIcon,
    FrSpinner,
    FrNavbar,
    BDropdown,
    BDropdownItem,
    BButton,
  },
  data() {
    return {
      firstLoad: true,
    };
  },
  props: {
    campaignDetails: {
      type: Object,
      default: () => {},
    },
    isComplete: {
      type: Boolean,
      default: false,
    },
    hideSignOff: {
      type: Boolean,
      default: false,
    },
    isSaving: {
      type: Boolean,
      default: null,
    },
  },
  computed: {
    disableSignoff() {
      // disable while saving
      if (this.isSaving) return true;

      // disable if campaign does not allow partial signoff and access review is not complete
      if (!this.campaignDetails.allowPartialSignoff && !this.isComplete) return true;

      return false;
    },
  },
  methods: {
    onForwardAllTasks() {
      this.$emit('change-saving');
      forwardCertificationTasks(this.campaignDetails.id)
        .catch((error) => {
          this.showErrorMessage(error, this.$t('governance.certificationTask.errors.onForwardError'));
        }).finally(() => {
          this.$emit('change-saving');
        });
    },
  },
  watch: {
    isSaving() {
      this.firstLoad = false;
    },
  },
};
</script>
