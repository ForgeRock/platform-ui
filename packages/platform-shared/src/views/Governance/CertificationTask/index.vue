<!-- Copyright (c) 2023-2024 ForgeRock. All rights reserved.

This software may be modified and distributed under the terms
of the MIT license. See the LICENSE file for details. -->
<template>
  <div class="certification-task-container">
    <FrCertificationToolbar
      :actor-id="actorId"
      :campaign-details="campaignDetails"
      :hide-sign-off="hideSignOff"
      :is-complete="isComplete"
      :is-saving="isSaving"
      :task-status="taskStatus"
      @change-saving="setSaving"
      @review-forwarded="goToBackUrl"
      @sign-off="signOff" />
    <p
      v-if="loadFailed"
      class="lead text-center m-5">
      {{ $t("pages.loadPage.errorLoadingPage") }}
    </p>
    <template v-else>
      <FrCertificationDetails
        :totals="totals"
        :campaign-details="campaignDetails"
        :is-loading="isDetailsLoading" />
      <div class="border-top position-relative">
        <template v-if="!isGroupByAccount">
          <BTabs
            :nav-class="['fr-tabs', 'pl-4', { 'd-none': hideTabs}]"
            data-testid="certification-tasklist-tabs"
            lazy>
            <BTab
              v-for="tab in grantTypeTabs"
              :key="tab.key"
              title-link-class="py-4 text-capitalize"
              :data-testid="`cert-${tab.key}-tab`"
              :title="tab.label">
              <FrTaskList
                v-if="campaignId && actorId"
                :certification-grant-type="tab.key"
                :campaign-id="campaignId"
                :campaign-details="campaignDetails"
                :refresh-tasks="refreshTasks"
                :is-admin="isAdmin"
                :actor-id="actorId"
                :task-status="taskStatus"
                @hide-group-by="hideGroupBy"
                @change-saving="setSaving"
                @check-progress="checkInProgress"
                @refresh-complete="refreshTasks = false"
                @signed-off="hideSignOff = true;"
                @set-totals="totals = $event"
                @update-details="getCertificationDetails" />
            </BTab>
          </BTabs>
        </template>
        <FrTaskListGroupBy
          v-else
          certification-grant-type="accounts"
          :campaign-id="campaignId"
          :campaign-details="campaignDetails"
          :refresh-tasks="refreshTasks"
          :is-admin="isAdmin"
          :actor-id="actorId"
          :show-group-by="isGroupByAccount"
          :task-status="taskStatus"
          @change-saving="setSaving"
          @check-progress="checkInProgress"
          @refresh-complete="refreshTasks = false"
          @signed-off="hideSignOff = true;"
          @set-totals="totals = $event"
          @update-details="getCertificationDetails"
        />
        <FrField
          v-if="showGroupByField"
          class="mb-4 text-capitalize group-by-position"
          v-model="isGroupByAccount"
          name="certificationGroupByAccount"
          testid="certification-group-by-account"
          type="checkbox"
          :label="$t('governance.certificationTask.certificationTabs.groupByAccount')" />
      </div>
    </template>
  </div>
</template>

<script>
import useBreadcrumb from '@forgerock/platform-shared/src/composables/breadcrumb';
import {
  BTabs,
  BTab,
} from 'bootstrap-vue';
import {
  getCertificationDetails,
  getInProgressTasksByCampaign,
  signOffCertificationTasks,
} from '@forgerock/platform-shared/src/api/governance/CertificationApi';
import NotificationMixin from '@forgerock/platform-shared/src/mixins/NotificationMixin';
import FrField from '@forgerock/platform-shared/src/components/Field';
import FrCertificationDetails from './TaskHeader/CertificationDetails';
import FrCertificationToolbar from './TaskHeader/CertificationToolbar';
import FrTaskList from './TaskList';
import FrTaskListGroupBy from './TaskListGroupBy';

export default {
  name: 'CertificationTask',
  mixins: [
    NotificationMixin,
  ],
  components: {
    BTab,
    BTabs,
    FrCertificationDetails,
    FrCertificationToolbar,
    FrField,
    FrTaskList,
    FrTaskListGroupBy,
  },
  props: {
    isAdmin: {
      type: Boolean,
      default: false,
    },
  },
  setup() {
    const { setBreadcrumb } = useBreadcrumb();
    return { setBreadcrumb };
  },
  data() {
    return {
      actorId: null,
      campaignDetails: {},
      campaignId: null,
      certifierId: null,
      hideSignOff: true,
      isDetailsLoading: false,
      isSaving: false,
      loadFailed: false,
      refreshTasks: false,
      taskStatus: null,
      totals: null,
      isGroupByAccount: false,
      showGroupByAccount: true,
      backUrl: '/access-reviews',
    };
  },
  computed: {
    hideTabs() {
      return this.grantTypeTabs.length === 1;
    },
    grantTypeTabs() {
      const tabs = [];
      if (this.isAccountTargetFilter) {
        tabs.push({
          key: 'accounts',
          label: this.$t('governance.certificationTask.certificationTabs.accounts'),
        });
      }
      if (this.isEntitlementTargetFilter) {
        tabs.push({
          key: 'entitlements',
          label: this.$t('governance.certificationTask.certificationTabs.entitlements'),
        });
      }
      if (this.isRoleTargetFilter) {
        tabs.push({
          key: 'roles',
          label: this.$t('common.roles'),
        });
      }
      return tabs;
    },
    isComplete() {
      return this.totals?.NONE === undefined;
    },
    isEntitlementCertificationType() {
      return this.campaignDetails.certificationType === 'entitlement';
    },
    isAccountTargetFilter() {
      return this.isGrantType('accountGrant');
    },
    isEntitlementTargetFilter() {
      return this.isGrantType('entitlementGrant');
    },
    isRoleTargetFilter() {
      return this.isGrantType('roleMembership');
    },
    showGroupByField() {
      return this.isAccountTargetFilter && this.isEntitlementTargetFilter && !this.isRoleTargetFilter && this.showGroupByAccount;
    },
  },
  methods: {
    getCertificationDetails() {
      this.isDetailsLoading = true;

      getCertificationDetails(this.campaignId)
        .then(({ data }) => {
          this.loadFailed = false;
          this.campaignDetails = data;
          this.actorId = this.$route.query.actorId;
        })
        .catch((error) => {
          this.loadFailed = true;
          this.showErrorMessage(error, this.$t('governance.certificationTask.errors.certificationDetailError'));
        })
        .finally(() => {
          this.isDetailsLoading = false;
        });
    },
    setSaving() {
      this.isSaving = !this.isSaving;
    },
    checkInProgress() {
      getInProgressTasksByCampaign(this.campaignId, this.isAdmin, this.taskStatus).then(({ data }) => {
        // verifies if the user has at least one sign-off permission for line items
        const atLeastOneSignoffPermission = (item) => {
          const reviewer = item.decision.certification.actors.find((actor) => actor.id === this.actorId);
          return reviewer?.permissions?.signoff;
        };
        const hasSignoffPermission = data.result.some(atLeastOneSignoffPermission);
        this.hideSignOff = !data.result.length || !hasSignoffPermission;
      });
    },
    signOff() {
      this.setSaving();
      signOffCertificationTasks(this.campaignDetails.id, this.actorId)
        .catch((error) => {
          this.showErrorMessage(error, this.$t('governance.certificationTask.errors.signOffError'));
        }).finally(() => {
          this.refreshTasks = true;
          this.setSaving();
          if (this.isComplete) {
            this.goToBackUrl();
          }
        });
    },
    goToBackUrl() {
      this.$router.push(this.backUrl);
    },
    hideGroupBy() {
      this.isGroupByAccount = false;
      this.showGroupByAccount = false;
    },
    /**
     * Verifies if a grant type is present in the campaign using the target filter types property.
     * @param {string} type - type of grant
     */
    isGrantType(type) {
      return this.campaignDetails.targetFilter?.type?.includes(type);
    },
  },
  mounted() {
    this.campaignId = this.$route?.params?.campaignId;
    this.actorId = this.$route.query.actorId;
    this.taskStatus = this.$route.query.taskStatus;
    this.getCertificationDetails();
    if (this.isAdmin) {
      this.backUrl = `/certification/campaigns/${this.campaignId}/access-reviews`;
    }
    this.setBreadcrumb(this.backUrl, this.$t('governance.certificationTask.backLink'));
  },
};
</script>
<style lang="scss" scoped>
  .certification-task-container {
    background-color: $white;
    min-height: 100vh;
  }
  .group-by-position {
    position: absolute;
    top: 25px;
    right: 45px;
  }
</style>
