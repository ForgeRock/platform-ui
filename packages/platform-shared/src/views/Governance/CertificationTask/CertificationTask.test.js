/**
 * Copyright (c) 2023-2024 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import { mount, flushPromises } from '@vue/test-utils';
import { setupTestPinia } from '@forgerock/platform-shared/src/utils/testPiniaHelpers';
import * as CertificationApi from '@forgerock/platform-shared/src/api/governance/CertificationApi';
import * as CommonsApi from '@forgerock/platform-shared/src/api/governance/CommonsApi';
import { findByTestId } from '@forgerock/platform-shared/src/utils/testHelpers';
import Notifications from '@kyvg/vue3-notification';
import CertificationTask from './index';

jest.mock('@forgerock/platform-shared/src/api/governance/CertificationApi');

describe('CertificationTask', () => {
  let wrapper;

  CommonsApi.getGlossarySchema = jest.fn().mockImplementation(() => Promise.resolve({ data: {} }));
  CommonsApi.getFilterSchema = jest.fn().mockImplementation(() => Promise.resolve({ data: {} }));
  CertificationApi.getCertificationUserFilter.mockImplementation(() => Promise.resolve({ data: {} }));
  CertificationApi.certifyItem.mockImplementation(() => Promise.resolve({ data: {} }));
  CertificationApi.certifyItems.mockImplementation(() => Promise.resolve({ data: {} }));
  CertificationApi.certifyAllItems.mockImplementation(() => Promise.resolve({ data: {} }));
  CertificationApi.exceptionItem.mockImplementation(() => Promise.resolve({ data: {} }));
  CertificationApi.exceptionItems.mockImplementation(() => Promise.resolve({ data: {} }));
  CertificationApi.exceptionAllItems.mockImplementation(() => Promise.resolve({ data: {} }));
  CertificationApi.forwardItem.mockImplementation(() => Promise.resolve({ data: {} }));
  CertificationApi.forwardItems.mockImplementation(() => Promise.resolve({ data: {} }));
  CertificationApi.forwardAllItems.mockImplementation(() => Promise.resolve({ data: {} }));
  CertificationApi.getEntitlementDetails.mockImplementation(() => Promise.resolve({ data: {} }));
  CertificationApi.getUserDetails.mockImplementation(() => Promise.resolve({ data: {} }));
  CertificationApi.getUserDetailsByType.mockImplementation(() => Promise.resolve({ data: {} }));
  CertificationApi.reassignItem.mockImplementation(() => Promise.resolve({ data: {} }));
  CertificationApi.resetItem.mockImplementation(() => Promise.resolve({ data: {} }));
  CertificationApi.resetAllItems.mockImplementation(() => Promise.resolve({ data: {} }));
  CertificationApi.revokeItems.mockImplementation(() => Promise.resolve({ data: {} }));
  CertificationApi.revokeAllItems.mockImplementation(() => Promise.resolve({ data: {} }));
  CertificationApi.revokeItem.mockImplementation(() => Promise.resolve({ data: {} }));
  CertificationApi.saveComment.mockImplementation(() => Promise.resolve({ data: {} }));
  CertificationApi.updateActors.mockImplementation(() => Promise.resolve({ data: {} }));
  CertificationApi.getCertificationApplicationFilter.mockImplementation(() => Promise.resolve({ data: {} }));
  CertificationApi.getCertificationCounts.mockImplementation(() => Promise.resolve({ data: {} }));
  CertificationApi.getCertificationTasksListByCampaign.mockImplementation(() => Promise.resolve({ data: {} }));
  CertificationApi.signOffCertificationTasks.mockImplementation(() => Promise.resolve({}));
  CertificationApi.getCertificationDetails.mockImplementation(() => Promise.resolve({ data: {} }));
  CertificationApi.getInProgressTasksByCampaign.mockImplementation(() => Promise.resolve({
    data: {
      result: [
        {
          id: 'a96de99c-c638-4bdd-84cb-5fb559225154',
          decision: {
            certification: {
              actors: [
                {
                  id: 'a96de99c-c638-4bdd-84cb-5fb559225153',
                  permissions: {
                    signoff: true,
                  },
                },
              ],
            },
          },
        },
        {
          id: 'a96de99c-c638-4bdd-84cb-5fb559225155',
          decision: {
            certification: {
              actors: [
                {
                  id: 'a96de99c-c638-4bdd-84cb-5fb559225153',
                  permissions: {
                    signoff: true,
                  },
                },
              ],
            },
          },
        },
        {
          id: 'a96de99c-c638-4bdd-84cb-5fb559225156',
          decision: {
            certification: {
              actors: [
                {
                  id: 'a96de99c-c638-4bdd-84cb-5fb559225157',
                  permissions: {
                    signoff: true,
                  },
                },
              ],
            },
          },
        },
      ],
    },
  }));

  function mountComponent() {
    setupTestPinia();
    return mount(CertificationTask, {
      global: {
        mocks: {
          $t: (t) => t,
          $route: {
            params: {
              campaignId: 'a96de99c-c638-4bdd-84cb-5fb559225152',
            },
            query: {
              actorId: 'a96de99c-c638-4bdd-84cb-5fb559225153',
              taskStatus: 'active',
            },
          },
          $router: { push: jest.fn() },
        },
        plugins: [Notifications],
        stubs: [
          'RouterLink',
          'D3PieChart',
          'FrCertificationTaskList',
        ],
      },
    });
  }

  describe('tests without data', () => {
    beforeEach(async () => {
      wrapper = mountComponent();

      await flushPromises();
    });

    it('component should load correctly', () => {
      expect(wrapper.vm.isDetailsLoading).toBe(false);
      expect(wrapper.vm.loadFailed).toBe(false);
      expect(wrapper.vm.campaignDetails).toEqual({});
      expect(wrapper.vm.actorId).toBe('a96de99c-c638-4bdd-84cb-5fb559225153');
      expect(wrapper.vm.campaignId).toBe('a96de99c-c638-4bdd-84cb-5fb559225152');
    });

    describe('checkInProgress method', () => {
      afterEach(() => {
        expect(CertificationApi.getInProgressTasksByCampaign).toHaveBeenCalledWith('a96de99c-c638-4bdd-84cb-5fb559225152', false, 'active');
      });

      it('checkInProgress method, two line items with signoff permission should show sign-off button', async () => {
        wrapper.vm.checkInProgress();

        await flushPromises();

        expect(wrapper.vm.hideSignOff).toBe(false);
        const signOffButton = findByTestId(wrapper, 'signoff-button');
        expect(signOffButton.exists()).toBe(true);
      });

      it('checkInProgress method, two line items, just one with signoff permission should show sign-off button', async () => {
        CertificationApi.getInProgressTasksByCampaign.mockImplementation(() => Promise.resolve({
          data: {
            result: [
              {
                id: 'a96de99c-c638-4bdd-84cb-5fb559225154',
                decision: {
                  certification: {
                    actors: [
                      {
                        id: 'a96de99c-c638-4bdd-84cb-5fb559225153',
                        permissions: {
                          signoff: true,
                        },
                      },
                    ],
                  },
                },
              },
              {
                id: 'a96de99c-c638-4bdd-84cb-5fb559225155',
                decision: {
                  certification: {
                    actors: [
                      {
                        id: 'a96de99c-c638-4bdd-84cb-5fb559225153',
                        permissions: {
                          signoff: false,
                        },
                      },
                    ],
                  },
                },
              },
              {
                id: 'a96de99c-c638-4bdd-84cb-5fb559225156',
                decision: {
                  certification: {
                    actors: [
                      {
                        id: 'a96de99c-c638-4bdd-84cb-5fb559225157',
                        permissions: {
                          signoff: true,
                        },
                      },
                    ],
                  },
                },
              },
            ],
          },
        }));

        wrapper.vm.checkInProgress();

        await flushPromises();

        expect(wrapper.vm.hideSignOff).toBe(false);
        const signOffButton = findByTestId(wrapper, 'signoff-button');
        expect(signOffButton.exists()).toBe(true);
      });

      it('checkInProgress method, two line items without signoff permission should hide sign-off button', async () => {
        CertificationApi.getInProgressTasksByCampaign.mockImplementation(() => Promise.resolve({
          data: {
            result: [
              {
                id: 'a96de99c-c638-4bdd-84cb-5fb559225154',
                decision: {
                  certification: {
                    actors: [
                      {
                        id: 'a96de99c-c638-4bdd-84cb-5fb559225153',
                        permissions: {
                          signoff: false,
                        },
                      },
                    ],
                  },
                },
              },
              {
                id: 'a96de99c-c638-4bdd-84cb-5fb559225155',
                decision: {
                  certification: {
                    actors: [
                      {
                        id: 'a96de99c-c638-4bdd-84cb-5fb559225153',
                        permissions: {
                          signoff: false,
                        },
                      },
                    ],
                  },
                },
              },
              {
                id: 'a96de99c-c638-4bdd-84cb-5fb559225156',
                decision: {
                  certification: {
                    actors: [
                      {
                        id: 'a96de99c-c638-4bdd-84cb-5fb559225157',
                        permissions: {
                          signoff: true,
                        },
                      },
                    ],
                  },
                },
              },
            ],
          },
        }));

        wrapper.vm.checkInProgress();

        await flushPromises();

        expect(wrapper.vm.hideSignOff).toBe(true);
        const signOffButton = findByTestId(wrapper, 'signoff-button');
        expect(signOffButton.exists()).toBe(false);
      });
    });
  });

  describe('tabs accounts and entitlement validation', () => {
    it('should load accounts and entitlement tabs because there is a target filter defined for both, group by account field shown', async () => {
      CertificationApi.getCertificationDetails.mockImplementation(() => Promise.resolve({
        data: {
          targetFilter: {
            type: [
              'accountGrant',
              'entitlementGrant',
            ],
          },
        },
      }));

      wrapper = mountComponent();

      await flushPromises();

      const accountsTab = findByTestId(wrapper, 'cert-accounts-tab');
      expect(accountsTab.exists()).toBe(true);

      const entitlementsTab = findByTestId(wrapper, 'cert-entitlements-tab');
      expect(entitlementsTab.exists()).toBe(true);

      const groupByAccountField = findByTestId(wrapper, 'certification-group-by-account');
      expect(groupByAccountField.exists()).toBe(true);
    });

    it('should load only accounts tab because there is not a target filter defined for entitlements, gruop by account field not shown', async () => {
      CertificationApi.getCertificationDetails.mockImplementation(() => Promise.resolve({
        data: {
          targetFilter: {
            type: [
              'accountGrant',
            ],
          },
        },
      }));

      wrapper = mountComponent();

      await flushPromises();

      const accountsTab = findByTestId(wrapper, 'cert-accounts-tab');
      expect(accountsTab.exists()).toBe(true);

      const entitlementsTab = findByTestId(wrapper, 'cert-entitlements-tab');
      expect(entitlementsTab.exists()).toBe(false);

      const groupByAccountField = findByTestId(wrapper, 'certification-group-by-account');
      expect(groupByAccountField.exists()).toBe(false);
    });

    it('should load only entitlements tab because there is not a target filter defined for accounts, gruop by account field not shown', async () => {
      CertificationApi.getCertificationDetails.mockImplementation(() => Promise.resolve({
        data: {
          targetFilter: {
            type: [
              'entitlementGrant',
            ],
          },
        },
      }));

      wrapper = mountComponent();

      await flushPromises();

      const accountsTab = findByTestId(wrapper, 'cert-accounts-tab');
      expect(accountsTab.exists()).toBe(false);

      const entitlementsTab = findByTestId(wrapper, 'cert-entitlements-tab');
      expect(entitlementsTab.exists()).toBe(true);

      const groupByAccountField = findByTestId(wrapper, 'certification-group-by-account');
      expect(groupByAccountField.exists()).toBe(false);
    });
  });

  describe('signOff', () => {
    it('should not go to back URL when cert is not complete', async () => {
      wrapper = mountComponent();
      await flushPromises();

      const goToBackUrlSpy = jest.spyOn(wrapper.vm, 'goToBackUrl');
      wrapper.vm.totals = {
        NONE: 1,
      };

      wrapper.vm.signOff();
      await flushPromises();
      expect(goToBackUrlSpy).not.toHaveBeenCalled();
    });
    it('should go to back URL when cert is complete', async () => {
      wrapper = mountComponent();
      await flushPromises();

      const goToBackUrlSpy = jest.spyOn(wrapper.vm, 'goToBackUrl');
      wrapper.vm.totals = {
        completed: 1,
      };

      wrapper.vm.signOff();

      await flushPromises();
      expect(goToBackUrlSpy).toHaveBeenCalled();
    });
  });
});
