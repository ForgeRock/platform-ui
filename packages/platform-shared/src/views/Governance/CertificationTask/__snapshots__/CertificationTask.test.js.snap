// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CertificationTask signOff should go to back URL when cert is complete: console-error-snapshot 1`] = `
[
  [
    "Un-mocked call made to Axios create. All API calls should be mocked",
  ],
  [
    "Un-mocked API call made with Axios. All API calls should be mocked",
  ],
  [
    "Un-mocked call made to Axios create. All API calls should be mocked",
  ],
  [
    "Un-mocked API call made with Axios. All API calls should be mocked",
  ],
]
`;

exports[`CertificationTask signOff should not go to back URL when cert is not complete: console-error-snapshot 1`] = `
[
  [
    "Un-mocked call made to Axios create. All API calls should be mocked",
  ],
  [
    "Un-mocked API call made with Axios. All API calls should be mocked",
  ],
  [
    "Un-mocked call made to Axios create. All API calls should be mocked",
  ],
  [
    "Un-mocked API call made with Axios. All API calls should be mocked",
  ],
  [
    "Un-mocked call made to Axios create. All API calls should be mocked",
  ],
  [
    "Un-mocked API call made with Axios. All API calls should be mocked",
  ],
  [
    "Un-mocked call made to Axios create. All API calls should be mocked",
  ],
  [
    "Un-mocked API call made with Axios. All API calls should be mocked",
  ],
]
`;

exports[`CertificationTask tabs accounts and entitlement validation should load accounts and entitlement tabs because there is a target filter defined for both, group by account field shown: console-error-snapshot 1`] = `
[
  [
    "Un-mocked call made to Axios create. All API calls should be mocked",
  ],
  [
    "Un-mocked API call made with Axios. All API calls should be mocked",
  ],
  [
    "Un-mocked call made to Axios create. All API calls should be mocked",
  ],
  [
    "Un-mocked API call made with Axios. All API calls should be mocked",
  ],
]
`;

exports[`CertificationTask tabs accounts and entitlement validation should load only accounts tab because there is not a target filter defined for entitlements, gruop by account field not shown: console-error-snapshot 1`] = `
[
  [
    "Un-mocked call made to Axios create. All API calls should be mocked",
  ],
  [
    "Un-mocked API call made with Axios. All API calls should be mocked",
  ],
  [
    "Un-mocked call made to Axios create. All API calls should be mocked",
  ],
  [
    "Un-mocked API call made with Axios. All API calls should be mocked",
  ],
  [
    "Un-mocked call made to Axios create. All API calls should be mocked",
  ],
  [
    "Un-mocked API call made with Axios. All API calls should be mocked",
  ],
  [
    "Un-mocked call made to Axios create. All API calls should be mocked",
  ],
  [
    "Un-mocked API call made with Axios. All API calls should be mocked",
  ],
]
`;

exports[`CertificationTask tabs accounts and entitlement validation should load only entitlements tab because there is not a target filter defined for accounts, gruop by account field not shown: console-error-snapshot 1`] = `
[
  [
    "Un-mocked call made to Axios create. All API calls should be mocked",
  ],
  [
    "Un-mocked API call made with Axios. All API calls should be mocked",
  ],
  [
    "Un-mocked call made to Axios create. All API calls should be mocked",
  ],
  [
    "Un-mocked API call made with Axios. All API calls should be mocked",
  ],
  [
    "Un-mocked call made to Axios create. All API calls should be mocked",
  ],
  [
    "Un-mocked API call made with Axios. All API calls should be mocked",
  ],
  [
    "Un-mocked call made to Axios create. All API calls should be mocked",
  ],
  [
    "Un-mocked API call made with Axios. All API calls should be mocked",
  ],
]
`;
